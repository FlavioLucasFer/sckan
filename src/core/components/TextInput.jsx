import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import $ from 'jquery';
import { uniqueId } from 'lodash';

import inputMaskApplier from 'core/utils/inputMaskApplier';
import inputValidator from 'core/utils/inputValidator';
import Icon from 'core/components/Icon';

TextInput.propTypes = {
	title: PropTypes.string,
	value: PropTypes.string,
	className: PropTypes.string,
	inputClassName: PropTypes.string,
	icon: PropTypes.string,
	iconClick: PropTypes.func,
	validate: PropTypes.bool,
	validateBy: PropTypes.string,
	mandatory: PropTypes.bool,
	mask: PropTypes.string,
	dataError: PropTypes.string,
	dataSuccess: PropTypes.string,
};

function TextInput(props) {
	const {
		title,
		value,
		className,
		inputClassName,
		icon,
		iconClick,
		validate,
		validateBy,
		mandatory,
		mask,
		dataError,
		dataSuccess,
		...rest
	} = props;

	const id = uniqueId('text-input-');
	const labelId = uniqueId('text-input-label-');
	const iconId = uniqueId('text-input-icon-');
	const helperTextId = uniqueId('text-input-helper-text-');

	const [firstFocus, setfirstFocus] = useState(false);
	const [helperMessage, setHelperMessage] = useState(mandatory ? '* This field is mandatory' : '');

	useEffect(() => {
		M.updateTextFields();
	}, []);

	useEffect(() => {
		if (validate && firstFocus) {
			if (!mandatory && (value && value.toString().trim() === ''));
			else {
				$(`#${id}`).removeClass('valid invalid');
				$(`#${id}`).addClass(isValid() ? 'valid' : 'invalid');
				$(`#${iconId}`).removeClass('green-text red-text teal-text');
				$(`#${iconId}`).addClass(isValid() ? 'green-text' : 'red-text');
				$(`#${helperTextId}`).removeClass('green-text red-text');
				$(`#${helperTextId}`).addClass(isValid() ? 'green-text' : 'red-text');
			}
		}
	});

	function isValid() {
		let isValid = true;
		let helperMessage = '';

		if (!validate)
			return true;

		else if (mandatory && (value && value.toString().trim() === '')) {
			helperMessage = '* This field is mandatory';
			isValid = false;
		}

		else if (validateBy) {
			const [valid, errorMessage] = inputValidator(validateBy, value);
			isValid = valid;
			helperMessage = errorMessage;
		} 

		setHelperMessage(helperMessage);
		return isValid;
	}

	function handleFocus() {
		if (!firstFocus) 
			setfirstFocus(true);
	
		$(`#${labelId}`).addClass('active');

		$(`#${iconId}`).removeClass('grey-text');
		$(`#${iconId}`).addClass(isValid() ? 'teal-text' : 'red-text');
	}
	
	function handleBlur() {
		if (!value) {
			$(`#${labelId}`).removeClass('active');
		}
		
		$(`#${iconId}`).removeClass(isValid() ? 'teal-text' : 'red-text');
		$(`#${iconId}`).addClass('grey-text');
	}

	function renderIcon() {
		if (!icon)
			return null;

		return (
			<Icon id={iconId}
				className={`prefix ${iconClick && 'pinter-cursor'}`}
				color="grey-text"
				onClick={iconClick && iconClick}>
				{icon}
			</Icon>
		);
	}

	function renderHelperMessage() {
		if (!helperMessage)
			return null;

		return (
			<span id={helperTextId}
				className="helper-text bold-text"
				data-error={dataError}
				data-success={dataSuccess}>
				{helperMessage}
			</span>
		);
	}

	return (
		<div className={`input-field ${className}`}>
			{renderIcon()}

			<input {...rest} 
				id={id}
				value={mask ? inputMaskApplier(mask, value) : value}
				className={inputClassName}
				onFocus={handleFocus}
				onBlur={handleBlur} />

			<label id={labelId} htmlFor={id}>{title}</label>

			{renderHelperMessage()}
		</div>
	);
}

export default TextInput;