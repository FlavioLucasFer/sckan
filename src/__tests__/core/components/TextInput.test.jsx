import React from 'react';
import { shallow } from 'enzyme';

import TextInput from 'core/components/TextInput';

describe('Testing TextInput core component', () => {
	it('should render correctly', () => {
		const wrapper = shallow(
			<TextInput />
		);

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.setProps({
			title: 'textinput test',
			value: 'test@mail.com',
			className: 'teal',
			inputClassName: 'grey-text',
			icon: 'visibility',
			validate: true,
			validateBy: 'email',
			mandatory: true,
			dataError: 'testing error',
			dataSuccess: 'testing success',
		});

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.setProps({
			value: '',
		});

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.find('input').simulate('focus');

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.setProps({
			value: '12312312312',
			validate: true,
			validateBy: 'cpf',
		});

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.setProps({
			value: '47057888007',
			validate: true,
			validateBy: 'cpf',
		});

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.setProps({
			value: '',
			validateBy: 'cpf',
			mask: 'cpf',
		});

		expect(wrapper.debug()).toMatchSnapshot();

		wrapper.find('input').simulate('change', '47057888007');

		expect(wrapper.debug()).toMatchSnapshot();
	});
});