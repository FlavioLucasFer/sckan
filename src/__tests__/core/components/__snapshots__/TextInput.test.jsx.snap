// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TextInput core component should render correctly 1`] = `
"<div className=\\"input-field undefined\\">
  <input id=\\"text-input-1\\" value={[undefined]} className={[undefined]} onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-2\\" htmlFor=\\"text-input-1\\" />
</div>"
`;

exports[`Testing TextInput core component should render correctly 2`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-7\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-5\\" value=\\"test@mail.com\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-6\\" htmlFor=\\"text-input-5\\">
    textinput test
  </label>
</div>"
`;

exports[`Testing TextInput core component should render correctly 3`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-11\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-9\\" value=\\"\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-10\\" htmlFor=\\"text-input-9\\">
    textinput test
  </label>
</div>"
`;

exports[`Testing TextInput core component should render correctly 4`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-19\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-17\\" value=\\"\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-18\\" htmlFor=\\"text-input-17\\">
    textinput test
  </label>
  <span id=\\"text-input-helper-text-20\\" className=\\"helper-text bold-text\\" data-error=\\"testing error\\" data-success=\\"testing success\\">
    * This field is mandatory
  </span>
</div>"
`;

exports[`Testing TextInput core component should render correctly 5`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-23\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-21\\" value=\\"12312312312\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-22\\" htmlFor=\\"text-input-21\\">
    textinput test
  </label>
  <span id=\\"text-input-helper-text-24\\" className=\\"helper-text bold-text\\" data-error=\\"testing error\\" data-success=\\"testing success\\">
    * This field is mandatory
  </span>
</div>"
`;

exports[`Testing TextInput core component should render correctly 6`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-27\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-25\\" value=\\"47057888007\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-26\\" htmlFor=\\"text-input-25\\">
    textinput test
  </label>
  <span id=\\"text-input-helper-text-28\\" className=\\"helper-text bold-text\\" data-error=\\"testing error\\" data-success=\\"testing success\\">
    * This field is mandatory
  </span>
</div>"
`;

exports[`Testing TextInput core component should render correctly 7`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-31\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-29\\" value=\\"\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-30\\" htmlFor=\\"text-input-29\\">
    textinput test
  </label>
  <span id=\\"text-input-helper-text-32\\" className=\\"helper-text bold-text\\" data-error=\\"testing error\\" data-success=\\"testing success\\">
    * This field is mandatory
  </span>
</div>"
`;

exports[`Testing TextInput core component should render correctly 8`] = `
"<div className=\\"input-field teal\\">
  <Icon id=\\"text-input-icon-31\\" className=\\"prefix undefined\\" color=\\"grey-text\\" onClick={[undefined]}>
    visibility
  </Icon>
  <input id=\\"text-input-29\\" value=\\"\\" className=\\"grey-text\\" onFocus={[Function: handleFocus]} onBlur={[Function: handleBlur]} />
  <label id=\\"text-input-label-30\\" htmlFor=\\"text-input-29\\">
    textinput test
  </label>
  <span id=\\"text-input-helper-text-32\\" className=\\"helper-text bold-text\\" data-error=\\"testing error\\" data-success=\\"testing success\\">
    * This field is mandatory
  </span>
</div>"
`;
